### ISSUE TITLE
Can't create webpack module with dependOn for import

### ISSUE THREAD

**Issue by lavcraft** [webpack/webpack#15616]
Hello. I would like to create three entry points with, on of them is shared.

I'm trying to use `dependOn` entry option. When i will have been added dependOn into entry, entry format changed. 

Run code like this in my frontend application:
```import * as X from 'module'); console.log(X);
```
And see for module [with **dependOn:**](https://github.com/lavcraft/react-hooks-composition-in-webpack-entry-problem-example/blob/master/packages/sdk/webpack.config.js#L24) :
![image](https://user-images.githubusercontent.com/576422/161279785-5df569e2-447f-4a58-b405-9ac865ad3cdd.png)

Also for module without **dependOn:** :
![image](https://user-images.githubusercontent.com/576422/161279350-77a32612-3c15-488d-9310-eb4d2bee9215.png)

I don't understand, how to right use `dependOn` for share some code between entris, because when i use it, i will see the import problem described above. Anyone could help, and explain how do it right? I can contribute into examples.

My reproducible example - https://github.com/lavcraft/react-hooks-composition-in-webpack-entry-problem-example

PS: I'm tried differente entry.library.type - all produce different kind of errors. Screenshots above made with [entry.library.type: 'this'](https://github.com/lavcraft/react-hooks-composition-in-webpack-entry-problem-example/blob/master/packages/sdk/webpack.config.js#L25) for debug purposes.

**Comment by lavcraft** [comment]
@vankop could it be a bug?
I think, in additional to `webpackChunk_company_sdk` this import should contain exports from this module (useDevices, withDevices in my example)

**Comment by vankop** [comment]
`dependOn` means that parent entry is loaded already, so in case:
```
entry A depends on entry B
```
when executing entry A all modules from entry B already loaded and presented in modules cache.

but I'm afraid this option does not work well with libraries, only for bundling application code.

**Comment by lavcraft** [comment]
@vankop thanks. It's interesting, may be somebody can explain this and prove - this  a bug or not. 

@sokra could you help me with this question?

**Comment by alexander-akait** [comment]
@lavcraft Why you need `dependOn` for libraries, I don't see any reasons for that

**Comment by lavcraft** [comment]
@alexander-akait thanks for question.
In a nutshell i demonstrate problem in [this example](https://github.com/lavcraft/react-hooks-composition-in-webpack-entry-problem-example).

Some details and history
1. I have a project with sdk and subprojects who use this sdk and shared project (shared between sdk, ui and backend). 
1.1. shared project included to `ui project` as builded dependency as a npm package. (For test purpose. Dogfood this package before distribute to external project)
1.2 shared project included to `backend and sdk projects` as typescript project reference

That's mean i need to build my sdk with tool, which support typescript and project references.
At first, i built this via rollup with some hacks and separated tsc process. But it's not reliable way and depended on build call order. Also it's **not** simply development process
In this configuration i have problem with typescript paths and aliases.

My next step - try other build systems for packaging sdk. I chose `webpack` , because ts-loader is good integrated with typescript and working good with project reference.

After build sdk with webpack, new problem will appear. This problem have a good explanation by @gaearon in [react issues](https://github.com/facebook/react/issues/24230#issuecomment-1084470436).
In short:
1. SDK must have a two different endpoints for avoiding load code, which not support SSR (external WebRTC sdk).
1.1 First endpoint contain `useCall()` hook with references to files which import external WebRTC packages.
1.2 Second endpoint contain `useDevices()` hook, which can be used in a lot of pages without full WebRTC call
2. But `useDevices` also required in `useCall()` for obtain current devices. And it's problem, because
```
withCall(withDevices(component)):
  useDevices() <-- return devices
   useCall()
     useDevices()  <-- return undefined, because React.useContext(DevicesContext) check references via equals, and in isolated webpack modules contain independent instance of this.
```

From this point i start trying to resolve this problem via `dependOn` and share logic between two endpoints. May there is a more suitable way to solve this, but i couldn't find it now :(

**Comment by gaearon** [comment]
Usually what you should do is extract the shared part into a separate bundle, and then use `externals` to prevent it from being copied into other bundles.

**Comment by lavcraft** [comment]
@gaearon right, i understand it. Thanks for this workaround. I prefer to see things through :) And i believe, if some tool have ability - it should works unless in docs not specified that it doesn't work :)

In this case, in my opinion, webpack should work with dependOn for library mode. It can simplify projects and provide ability to share code between entrypoints. Believe that webpack can be a best tool for build libraries and ready to try fix some problems with it

**Comment by lavcraft** [comment]
@gaearon @alexander-akait tried this way. But nothing changed.

Push reproducible results into [branch depends-on-experiments](https://github.com/lavcraft/react-hooks-composition-in-webpack-entry-problem-example/tree/depends-on-experiments)

Also comment in this thread - https://github.com/facebook/react/issues/24230#issuecomment-1086978269
